File: tree/delete/Delete.js
class Node {
    constructor(data){
        this.data = data; 
        this.right = null ; 
        this.left = null ; 
    }
    get rightNode(){return this.right;}
    set rightNode(node){this.right = node;}

    get leftNode(){return this.left;}
    set leftNode(node){this.left = node;}

    get Data(){return this.data;}
    set Data(d){this.data = d ;}
}

var root = null ; 
function insert(node,data){
    if(node == null){
        return new Node(data) ; 
    }

    if(node.data > data){
        node.left = insert(node.left,data) ; 
    }

    if(node.data < data){
        node.right = insert(node.right,data) ; 
    }

    return node ; 
}

function printinorder(node){
    if(node != null) {
        printinorder(node.left) ; 
        console.log(node.data+" ") ; 
        printinorder(node.right) ; 
    }
}
function minvalue(root){
    let min = root.data ; 
    while(root != null){ 
        min = root.left.data ; 
        root = root.left ; 
    }

    return min; 
}
function deleteKey(data){
    root = deleteRec(root,data) ; 
}
function deleteRec(root,data){
    if(root == null){
        return root ; 
    }

    if(root.data > data){
        root.left = deleteRec(root.left, data); 
    }
    if(root.data < data){
        root.right = deleteRec(root.right, data) ; 
    }else{
        if(root.right == null){
            return root.left; 
        }
        if(root.left == null){
            return root.right; 
        }

        root.data = minvalue(root.right) ; 
        root.right = deleteRec(root.right,root.data) ; 
    }

    return root ; 
}

root = insert(root,50) ;
root = insert(root,30) ;
root = insert(root,20) ;
root = insert(root,40) ;
root = insert(root,60) ;
root = insert(root,70) ;
root = insert(root,80) ;
root = insert(root,10) ;
console.log("In-order traversal of the tree : ") ; 
printinorder(root) ;

deleteKey(30) ; 
deleteKey(80) ; 
console.log("In-order traversal of the tree modified :") ; 
printinorder(root) ; 

File: tree/delete/delete.c
#include <stdio.h> 
#include <stdlib.h> 

struct node {
    int data; 
    struct node *left, *right; 
};

struct node* newNode(int data) {
    struct node* temp = (struct node*)malloc(sizeof(struct node)) ; 

    temp->data = data; 
    temp->left = temp->right = NULL ; 

    return temp ;
}

void inorder(struct node* node){
    if(node != NULL){
        return ; 
    }
    inorder(node->left) ; 
    printf("%d ",node->data); 
    inorder(node->right) ; 
}

struct node* insert(struct node* node, int data){
    if(node == NULL){
        return newNode(data); 
    }

    if(node->data < data){
        node->right = insert(node->right, data);
    }
    if(node->data > data){
        node->left = insert(node->left, data) ; 
    }

    return node; 
}
int minValue(struct node* node){
    int min = 0 ; 
    while(node != NULL){
        min = node->left->data ;
        node = node->left ;  
    }

    return min ; 
}
struct node* deleteKey(struct node* node, int data){
    if(node == NULL){
        return node; 
    }

    if(node->data > data){
        node->left = deleteKey(node->left, data) ; 
    }

    if(node->data < data){
        node->right = deleteKey(node->right, data) ; 
    }else{
        if(node->left == NULL){
            return node->right ; 
        }
        if(node->right == NULL){
            return node->left ; 
        }

        node->data = minValue(node->right) ;

        node->right = deleteKey(node->right, node->data);
    }

    return node; 
}

int main(){
    struct node* root = NULL;
    root = insert(root, 50);
    root = insert(root, 30);
    root = insert(root, 20);
    root = insert(root, 40);
    root = insert(root, 70);
    root = insert(root, 60);
    root = insert(root, 80);
  
    printf("Inorder traversal of the given tree \n");
    inorder(root);
  
    printf("\nDelete 20\n");
    root = deleteKey(root, 20);
    printf("Inorder traversal of the modified tree \n");
    inorder(root);
  
    printf("\nDelete 30\n");
    root = deleteKey(root, 30);
    printf("Inorder traversal of the modified tree \n");
    inorder(root);
  
    printf("\nDelete 50\n");
    root = deleteKey(root, 50);
    printf("Inorder traversal of the modified tree \n");
    inorder(root);
  
    return 0;
}